pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractReplicator.java

```diff
@@ -188,58 +270,135 @@ protected CompletableFuture<Boolean> isLocalTopicActive() {
+                            + " state : {}", replicatorId, state);
             return CompletableFuture.completedFuture(null);
         }
-        CompletableFuture<Void> future = producer.closeAsync();
+
+        // Close producer and update state.
+        return doCloseProducerAsync(producer, () -> {
+            Pair<Boolean, State> setDisconnectedRes = compareSetAndGetState(State.Disconnecting, State.Disconnected);
+            if (setDisconnectedRes.getLeft()) {
+                this.producer = null;
+                // deactivate further read
+                disableReplicatorRead();
+                return;
+            }
+            if (setDisconnectedRes.getRight() == State.Terminating
+                    || setDisconnectingRes.getRight() == State.Terminated) {
+                log.info("[{}] Skip setting state to terminated because it was terminated, state : {}",
+                        replicatorId, state);
+            } else {
+                // Since only one task can call "doCloseProducerAsync(producer, action)", this scenario is not expected.
+                // So print a warn log.
+                log.warn("[{}] Other task has change the state to terminated. so skipped current one task."
+                                + " State is : {}",
+                        replicatorId, state);
+            }
+        });
+    }
+
+    protected CompletableFuture<Void> doCloseProducerAsync(Producer<byte[]> producer, Runnable actionAfterClosed) {
+        CompletableFuture<Void> future =
+                producer == null ? CompletableFuture.completedFuture(null) : producer.closeAsync();
         return future.thenRun(() -> {
-            STATE_UPDATER.set(this, State.Stopped);
-            this.producer = null;
-            // deactivate further read
-            disableReplicatorRead();
+            actionAfterClosed.run();
         }).exceptionally(ex -> {
             long waitTimeMs = backOff.next();
             log.warn(
-                    "[{}] Exception: '{}' occurred while trying to close the producer."
-                            + " retrying again in {} s",
-                    replicatorId, ex.getMessage(), waitTimeMs / 1000.0);
+                    "[{}] Exception: '{}' occurred while trying to close the producer. Replicator state: {}."
+                            + " Retrying again in {} s.",
+                    replicatorId, ex.getMessage(), state, waitTimeMs / 1000.0);
             // BackOff before retrying
-            brokerService.executor().schedule(this::closeProducerAsync, waitTimeMs, TimeUnit.MILLISECONDS);
+            brokerService.executor().schedule(() -> doCloseProducerAsync(producer, actionAfterClosed),
+                    waitTimeMs, TimeUnit.MILLISECONDS);
             return null;
         });
     }
 
+    public CompletableFuture<Void> terminate() {
+        if (!tryChangeStatusToTerminating()) {
```

Comment:
> The method tryChangeStatusToTerminating checks four states, Starting, Started, Disconnecting, Disconnected, maybe we can only check the states Terminating and Terminated here.

Since we need `compareAndSet`, so we need four logic branches. And the state `Disconnecting` and `Disconnected` will be removed in the future, the current method is not complex.

> Another potential problem is that if the state is Terminating, the method terminate will return a complete future object, but the terminate operation may not finish. 

The method `terminate` will always return a completed future, because we watched the exception by `exceptionally` in the method `doCloseProducerAsync`, and added a mechanism to guarantee the producer can be closed eventually.

Reply:
Got it, thanks.